tests/xtriggers/02-persistence.t

```diff
@@ -0,0 +1,62 @@
+#!/bin/bash
+# THIS FILE IS PART OF THE CYLC SUITE ENGINE.
+# Copyright (C) 2008-2017 NIWA
+# 
+# This program is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation, either version 3 of the License, or
+# (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+#------------------------------------------------------------------------------
+
+# Test persistence of xtrigger results across restart. A cycling task depends
+# on a non cycle-point dependent custom xtrigger called "faker". In the first
+# cycle point the xtrigger succeeds and returns a result, then a task shuts
+# the suite down.  Then we replace the custom xtrigger function with one that
+# will fail if called again - which should not happen because the original
+# result should be remembered (as this xtrigger is not cycle point dependent).
+# Also test the correct result is broadcast to the dependent task before and
+# after suite restart.
+
+. $(dirname $0)/test_header
+set_test_number 6
+
+install_suite ${TEST_NAME_BASE} ${TEST_NAME_BASE}
+
+# Install the succeeding xtrigger function.
+cp ${TEST_SOURCE_DIR}/${TEST_NAME_BASE}/faker-succ.py faker.py
+export PYTHONPATH=$PWD:$PYTHONPATH
+
+# Validate the test suite.
+run_ok ${TEST_NAME_BASE}-val cylc val --debug $SUITE_NAME
+
+# Run the first cycle, till auto shutdown by task.
+TEST_NAME=${TEST_NAME_BASE}-run
+suite_run_ok ${TEST_NAME} cylc run --no-detach --debug $SUITE_NAME
+
+# Check the broadcast result of xtrigger.
+cylc cat-log -o ${SUITE_NAME} foo.2010 > foo.2010.out
+grep_ok "NAME is bob" foo.2010.out
+
+# Replace the xtrigger function with one that will fail if called again.
+cp ${TEST_SOURCE_DIR}/${TEST_NAME_BASE}/faker-fail.py faker.py
+
+# Validate again (with the new xtrigger function).
+run_ok ${TEST_NAME_BASE}-val2 cylc val --debug $SUITE_NAME
+
+# Restart the suite, to run the final cycle point.
+TEST_NAME=${TEST_NAME_BASE}-restart
+suite_run_ok ${TEST_NAME} cylc restart --no-detach $SUITE_NAME
```

Comment:
This test is failing in my environment with this traceback in the suite's err file:

```
2017-11-17T11:06:02Z CRITICAL - Traceback (most recent call last):
	  File "/home/matt/cylc.git/lib/cylc/scheduler.py", line 235, in start
	    self.configure()
	  File "/home/matt/cylc.git/lib/cylc/scheduler.py", line 382, in configure
	    self.load_tasks_for_restart()
	  File "/home/matt/cylc.git/lib/cylc/scheduler.py", line 460, in load_tasks_for_restart
	    self.xtrigger_mgr.load_xtrigger_for_restart)
	  File "/home/matt/cylc.git/lib/cylc/rundb.py", line 639, in select_xtriggers_for_restart
	    callback(row_idx, list(row))
	  File "/home/matt/cylc.git/lib/cylc/xtrigger_mgr.py", line 123, in load_xtrigger_for_restart
	    self.sat_xtrig[sig] = pickle.loads(results_pickle)
	  File "/usr/lib64/python2.6/pickle.py", line 1374, in loads
	    return Unpickler(file).load()
	  File "/usr/lib64/python2.6/pickle.py", line 858, in load
	    dispatch[key](self)
	KeyError: '\x00'
```

Reply:
Huh, OK.

Reply:
(hopefully this is fine now, using json instead of pickle)